src/common/utils/fold.h

```diff
@@ -0,0 +1,57 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#ifndef SRC_COMMON_UTILS_FOLD_H_
+#define SRC_COMMON_UTILS_FOLD_H_
+
+// This file contains a C++ implementation of the classic "fold" and
+// "right_fold" higher-order functions. This is somewhat redundant with
+// std::accumulate; unfortunately, std::accumulate requires its accumulator to
+// be copyable until C++20.
+//
+// Note that "right_fold" is named "rfold". This is a bit of a pun: if you come
+// from a functional languages background, the "r" naturally expands to "right",
+// while if you are fluent with C++ iterators, "r" naturally means "reverse".
+// In this context it actually means both: a right_fold achieved by using
+// reverse iterators on the provided container.
+
+#include <iterator>
+
+namespace raksha::common::utils {
+
+template <class I, class A, class F>
+A fold_iter(I begin_iter, I end_iter, A accumulated, F fold_fn) {
+  if (begin_iter == end_iter) return accumulated;
+  const auto &current_element = *begin_iter;
+  return fold_iter(++begin_iter, end_iter,
```

Comment:
Can't you convert this to loop, if you change call in `fold` to pass reverse range and `rfold` to pass forward range?

(This suggestion makes sense only if `fold_iter` is private. Otherwise, let us leave this as is.)

Reply:
`fold_iter` is public, leaving as is.